app/Main.hs

```diff
@@ -0,0 +1,158 @@
+-- My lsp for some reason prefers this when formatting
+{-# LANGUAGE ImportQualifiedPost #-}
+{-# LANGUAGE OverloadedStrings #-}
+
+module Main (main) where
+
+import Control.Applicative (liftA2, (<|>))
+import Control.Monad (join, (<=<))
+import Data.Either.Extra (maybeToEither)
+import Data.Foldable (foldl')
+import Data.List (foldl1')
+import Data.Map (Map, alter, empty, (!?))
+import Data.Text qualified as Text
+import Data.Text.IO qualified as TextIO
+import System.Console.ANSI (Color (..), ColorIntensity (..), ConsoleLayer (..), SGR (..), clearScreen, setCursorPosition, setSGR)
+import Text.Read (readMaybe)
+
+type Coordinates = (Int, Int)
+
+data Player = Cross | Circle deriving (Eq, Show)
+
+showPlayerT :: Player -> Text.Text
+showPlayerT = Text.pack . show
+
+newtype Spot = Spot (Maybe Player)
+
+instance Show Spot where
+  show = Text.unpack . showSpotT
+
+showSpotT :: Spot -> Text.Text
+showSpotT (Spot (Just Cross)) = "x"
+showSpotT (Spot (Just Circle)) = "0"
+showSpotT (Spot Nothing) = " "
+
+newtype Board = Board (Map Coordinates Player)
+
+instance Show Board where
+  show = Text.unpack . showBoard
+
+showBoard :: Board -> Text.Text
+showBoard (Board board) = foldl' replaceExclamation unfilledBoard $ Spot . (board !?) <$> coordinates
+  where
+    -- Not Sure what I think about this implementations but it works and is pretty clean
+    coordinates :: [Coordinates]
+    coordinates = [(x, y) | y <- [0 .. 2], x <- [0 .. 2]]
+    replaceExclamation :: Text.Text -> Spot -> Text.Text
+    replaceExclamation boardText spot = replaceOne "!" (showSpotT spot) boardText
+
+    unfilledBoard =
+      "  0   1   2 \n\
+      \0 ! | ! | ! \n\
+      \ ---+---+---\n\
+      \1 ! | ! | ! \n\
+      \ ---+---+---\n\
+      \2 ! | ! | ! "
+
+nextTurn :: Player -> Player
+nextTurn Cross = Circle
+nextTurn Circle = Cross
+
+replaceOne :: Text.Text -> Text.Text -> Text.Text -> Text.Text
+replaceOne pattern substitution text = Text.concat [front, substitution, Text.drop (Text.length pattern) back]
+  where
+    (front, back) = Text.breakOn pattern text
+
+playAtSpot :: Coordinates -> Player -> Board -> Maybe Board
+playAtSpot coord spot (Board board) = case board !? coord of
+  Nothing -> Just $ Board $ alter (const $ Just spot) coord board
+  Just _ -> Nothing
+
+allEqMaybe :: Eq a => [a] -> Maybe a
+allEqMaybe [] = Nothing
+allEqMaybe xs | all (== head xs) xs = Just $ head xs
+allEqMaybe _ = Nothing
+
+winner :: Board -> Coordinates -> Maybe Player
+winner (Board board) (x, y) =
+  foldl1' (<|>) $ isWin <$> winningChecks
+  where
+    winningChecks
+      -- True if coordinates aren't part of diagonal
+      | (x == 1 || y == 1) && (x /= y) = [row, column]
+      | x == 1 && y == 1 = [row, column, diag2, diag1]
+      -- True for diag 1
+      | x == y = [row, column, diag1]
+      -- True for diag 2
+      | otherwise = [row, column, diag2]
+
+    isWin = join . allEqMaybe . ((board !?) <$>)
```

Comment:
`join . f . g` simplifies to i.e. `f <=< g`

Reply:
Here is doesn't.

Reply:
Maybe Monad transformers would fix this but I don't really know

Reply:
Actually `sequenceA` does give the desired behaviour it does the operation of ignoring triplets that have a spot that hasn't been played yet

Reply:
Linter told me to use `traverse f` insteade of `seqeunceA . (f <$>) `

Reply:
> Here is doesn't ...

My point is that `join . fmap f . g` is just bad style for i.e. `f <=< g`.
More generally, plumbing like `fmap`, `bind`, `join` and so on are (almost always) not the interesting part - it's just the price of encoding effects in the type system. That's the reason we have many different styles - operators, named functions, syntactic sugars (i.e. `do` notations, comprehensions).


> Linter told me to use `traverse f` instead of `seqeunceA . (f <$>) `

Yes, `traverse f` is basically `sequenceA . fmap f`.
This relates to the above point - an `fmap` followed immediately by some variant of `sequence` is just two consecutive typing acrobatics that can be easily merged into the right one (which is in the standard library, not some exotic behavior you'd need to name and may be tricky to get right).